src/ray/gcs/gcs_server/gcs_placement_group_manager.h

```diff
@@ -82,17 +89,72 @@ class GcsPlacementGroup {
   /// Get the Strategy
   rpc::PlacementStrategy GetStrategy() const;
 
-  // Get debug string for the placement group.
+  /// Get debug string for the placement group.
   std::string DebugString() const;
 
+  /// Below fields are used for automatic cleanup of placement groups.
+
+  /// Get the actor id that created the placement group.
+  const ActorID GetCreatorActorId() const;
+
+  /// Get the job id that created the placement group.
+  const JobID GetCreatorJobId() const;
+
+  /// Mark that the creator job of this placement group is dead.
+  void MarkCreatorJobDead();
+
+  /// Mark that the creator actor of this placement group is dead.
+  void MarkCreatorActorDead();
+
+  /// Return True if the placement group is removable. False otherwise.
+  bool IsPlacementGroupRemovable() const;
+
  private:
   /// The placement_group meta data which contains the task specification as well as the
   /// state of the gcs placement_group and so on (see gcs.proto).
   rpc::PlacementGroupTableData placement_group_table_data_;
 };
 
-using RegisterPlacementGroupCallback =
-    std::function<void(std::shared_ptr<GcsPlacementGroup>)>;
+using PlacementGroupMap =
+    absl::flat_hash_map<PlacementGroupID, std::shared_ptr<GcsPlacementGroup>>;
+
+/// An index to obtain GcsPlacementGroup from multiple keys.
+class RegisteredPlacementGroupIndex {
+ public:
+  explicit RegisteredPlacementGroupIndex() {}
+  ~RegisteredPlacementGroupIndex() = default;
+
+  /// Emplace the index of the placement of a given id.
+  bool Emplace(const PlacementGroupID &placement_group_id,
+               std::shared_ptr<GcsPlacementGroup> placement_group);
+
+  /// Erase the index of the placement group of a given id.
+  void Erase(const PlacementGroupID &placement_group_id);
+
+  /// Get the placement group of the corresponding id.
+  absl::optional<std::shared_ptr<GcsPlacementGroup>> Get(
+      const PlacementGroupID &placement_group_id) const;
+
+  /// Return the map of placement groups owned by a given actor id.
+  const PlacementGroupMap &GetPlacementGroupsOwnedByActor(const ActorID &actor_id);
+
+  /// Return the map of placement groups owned by a given job id.
+  const PlacementGroupMap &GetPlacementGroupsOwnedByJob(const JobID &job_id);
+
+  /// Return the map of registered placement groups. This is used to iterate
+  /// registered placement groups.
+  const PlacementGroupMap &GetRegisteredPlacementGroups() const;
+
+ private:
+  /// Map of the registered placement groups.
+  PlacementGroupMap placement_groups_;
+
+  /// Map of placement groups owned by actor ids.
+  absl::flat_hash_map<ActorID, PlacementGroupMap> placement_groups_by_actor_id_;
```

Comment:
Since the number of placement groups is not large (<1000), why not instead scan the list instead of keeping these auxiliary data structures?

Reply:
Hmm yeah. that was the original implementation actually. I thought it'd be better having this approach because gcs server is pretty busy already (by heartbeat), and single-threaded. I don't have a strong objection, but since I already wrote this data structure, and it is pretty well tested, why don't we just keep it?

Reply:
Well, a good reason to remove this is to keep the code unsurprising--- someone reading this in the future might get the impression this is needed for performance, and if happens to not be important they will probably be confused about why the design was this way.

I'm not too worried about GCS CPU time, it's not latency critical (out of critical path of most operations), and scans of small lists are infrequent and unlikely to ever be a bottleneck.